sources/api/datastore/src/filesystem.rs

```diff
@@ -489,6 +498,95 @@ impl DataStore for FilesystemDataStore {
         let pending = Committed::Pending {
             tx: transaction.into(),
         };
+
+        // We will first commit metadata to correctly check that if a setting exist and
+        // its strength is strong, we will not change it to weak.
+        // Get metadata from pending transaction
+        let transaction_metadata =
+            self.get_metadata_prefix("settings.", &pending, &None as &Option<&str>)?;
+
+        trace!(
+            "commit_transaction: transaction_metadata: {:?}",
+            transaction_metadata
+        );
+
+        for (key, value) in transaction_metadata {
+            for (metadata_key, metadata_value) in value {
+                // For now we are only processing the strength metadata from pending
+                // transaction to live
+                if metadata_key.name() != "strength" {
+                    continue;
+                }
```

Comment:
I chatted with @vyaghras about this in person and we had an idea -- we could modify `commit_transaction()` to support taking a function pointer to a commit "constraint check."

e.g.

```rust
pub struct ApprovedWrite {
    // Give the constraint check the ability to modify the write
    settings: Vec<(Key, Value)>,
    metadata: Vec<(Key, Key, Value)>,
}

pub enum ConstraintCheckResult {
    Reject,
    Approve(ApprovedWrite),
}

// Probably a good idea to implement these
impl From<ConstraintCheckResult> for Option<ApprovedWrite> {}
impl From<Option<ApprovedWrite>> for ConstraintCheckResult {}


fn commit_transaction<S>(&mut self, transaction: S, constraint_checK: &C) -> Result<HashSet<Key>>
where
    S: Into<String> + AsRef<str>,
    C: Fn(&impl Datastore, &Committed) -> Result<ConstraintCheckResult>;
```

Pros:
* This would allow the datastore to own things like reader/writer locks and the time at which the constraint check is performed. The caller has the ability to add constraints relating to the actual modeled data.
* Ability to add more model-specific constraints outside of the datastore implementation itself

Cons:
* Any callers would need to pass the right constraint check function, or you could have an avenue to commit transactions that don't meet our constraints.

The con could be worked around with generics. You could make the `Datastore` trait generic over a constraint check:

```rust
trait Datastore<Constraints>
where
    Constraints: Fn(&Datastore<Constraints>, &Committed) -> ConstraintCheckResult
```

Then we could do something like 
```rust
type BottlerocketDatastore = FilesystemDatastore<bottlerocket_commit_constraints>`
```

This adds a fair amount of complexity though. I'm open to keeping to one of the simpler solutions we've chatted about above.